databuilder/query_model.py

```diff
@@ -505,6 +507,23 @@ def get_input_nodes_for_categorise(node):
     return inputs
 
 
+def all_nodes(tree):
+    nodes = []
+
+    for subnode in get_input_nodes(tree):
+        for node in all_nodes(subnode):
+            nodes.append(node)
+    return [tree] + nodes
+
+
+def count_nodes(tree):  # pragma: no cover
+    return len(all_nodes(tree))
```

Comment:
I'm not sure this is accurate because the `all_nodes()` func it uses doesn't suppress duplicates.

Reply:
I don't understand this. Where do duplicates arise?

Reply:
Anytime there's more than one path to a given node. e.g. a Filter operation will always create two paths to its base SelectTable node: one via the `source` and one via the `condition`.

Reply:
I mean `count_nodes()` still measures _something_ related to the complexity of the query here, and maybe that's all that matters. But it doesn't literally count the nodes.

Reply:
These duplicates are equal but non-identical Python objects? I think I don't consider them duplicates in the query model (although they clearly point to the same conceptual _thing_ in the data).

(In fact this is another whole area for later discussion -- I have been secretly wondering if we would produce more efficient SQL if we de-duplicated these by turning the tree into a graph.)

Reply:
No, I think these will be the same Python objects appearing multiple times. (Agree about the de-duplication though.)

Reply:
Just to save going back and forth on this:
```
In [1]: from databuilder.query_model import *

In [2]: t = SelectTable("foo")

In [3]: f = Filter(t, Function.EQ(SelectColumn(t, "bar"), Value(10)))

In [4]: [id(i) for i in all_nodes(f)]
Out[4]: 
[140582029858608,
 140582046186800,
 140582035503136,
 140581660203136,
 140582046186800,
 140581660204960]
```

Reply:
Oh, this _is_ interesting. Whenever we write QM long-hand they are different objects. The generative tests (I'm pretty sure) are generating different objects. So you're talking about erhQL-generated QM?

This seems like maybe-a-problem if our tests are creating different types of data structures to our production code.

And maybe a bit weird that we have graphs which the engine is "reading" as trees.

Reply:
Ohhh, I see where you're coming from now. Well I don't think it's going to be a problem in practice because the frozen dataclasses make equivalent objects behave as pretty much identical unless you start using `is` or `id()`, which I don't think we do. But it's worth bearing in mind.

Reply:
I think we're agreed that this is vaguely interesting, but doesn't require changes now, is that right?

Reply:
Yes, agreed. That function does the right thing in the context of the generated tests.